pallets/snarcos/src/systems.rs

```diff
@@ -1,12 +1,67 @@
 use ark_poly::polynomial::univariate::DensePolynomial;
 use ark_poly_commit::marlin_pc::MarlinKZG10;
+use ark_relations::r1cs::SynthesisError;
 use ark_serialize::CanonicalDeserialize;
 use ark_snark::SNARK;
 use ark_std::rand::{prelude::StdRng, SeedableRng};
 use blake2::Blake2s;
 use codec::{Decode, Encode};
+use frame_support::{
+    log::{error, info},
+    PalletError,
+};
 use scale_info::TypeInfo;
 
+/// Possible errors from the verification process.
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Decode, Encode, TypeInfo, PalletError)]
+pub enum VerificationError {
+    /// The verifying key was malformed.
+    ///
+    /// May occur only for some non-universal system.
+    MalformedVerifyingKey,
+    /// There was an error in the underlying holographic IOP. For details, consult your logs.
+    ///
+    /// May occur only for some universal system.
+    AHPError,
+    /// There was an error in the underlying polynomial commitment. For details, consult your logs.
+    ///
+    /// May occur only for some universal system.
+    PolynomialCommitmentError,
+    /// Unexpected error has occurred. Check your logs.
+    UnexpectedError,
+}
+
+impl From<SynthesisError> for VerificationError {
+    fn from(syn_err: SynthesisError) -> Self {
+        match syn_err {
+            SynthesisError::MalformedVerifyingKey => VerificationError::MalformedVerifyingKey,
+            _ => {
+                error!("Unexpected SynthesisError variant: {syn_err}");
+                VerificationError::UnexpectedError
+            }
+        }
+    }
+}
+
+impl From<ark_marlin::Error<ark_poly_commit::Error>> for VerificationError {
+    fn from(err: ark_marlin::Error<ark_poly_commit::Error>) -> Self {
+        match err {
+            ark_marlin::Error::AHPError(err) => {
+                info!("Encountered AHP error: {err:?}");
```

Comment:
Is there a reason why this is logged as `info` rather than an error? Same for the next case.

Reply:
yes - in that case, we will still push the 'error status' back to the caller, which can be then handled. so in general, the logs shouldn't indicate that there's a program-level error

we are using info channel just to not miss the inner message (`AHPError`) - we cannot pass it to the caller due to unsatisfied traits, so it is the last place when the user can see it

Reply:
OK, gotcha